cloud/scope/managed_control_plane.go

```diff
@@ -669,6 +671,174 @@ func (s *ManagedControlPlaneScope) getSpecFromActual(cluster *oke.Cluster) *infr
+	if !(addon.LifecycleState == oke.AddonLifecycleStateUpdating ||
+		addon.LifecycleState == oke.AddonLifecycleStateFailed) {
+		addonConfigurationsActual := getActualAddonConfigurations(addon.Configurations)
+		// if the version changed or the configuration changed, update the addon
+		// if the lifecycle state is needs attention, try to update
+		if addon.LifecycleState == oke.AddonLifecycleStateNeedsAttention ||
+			!reflect.DeepEqual(addonInSpec.Version, addon.Version) ||
+			!reflect.DeepEqual(addonConfigurationsActual, addonInSpec.Configurations) {
+			s.Info(fmt.Sprintf("Updating addon %s", *addon.Name))
+			_, err := s.ContainerEngineClient.UpdateAddon(ctx, oke.UpdateAddonRequest{
+				ClusterId: okeCluster.Id,
+				AddonName: addon.Name,
+				UpdateAddonDetails: oke.UpdateAddonDetails{
+					Version:        addonInSpec.Version,
+					Configurations: getAddonConfigurations(addonInSpec.Configurations),
+				},
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *ManagedControlPlaneScope) handleDeletedAddon(ctx context.Context, okeCluster *oke.Cluster, addonName string) error {
+	resp, err := s.ContainerEngineClient.GetAddon(ctx, oke.GetAddonRequest{
+		ClusterId: okeCluster.Id,
+		AddonName: common.String(addonName),
+	})
+	if err != nil {
+		if ociutil.IsNotFound(err) {
+			s.OCIManagedControlPlane.RemoveAddonStatus(addonName)
+		} else {
+			return err
+		}
+	}
+	addonState := resp.LifecycleState
+	// if the addon is not in deleting state, call delete
+	if !(addonState == oke.AddonLifecycleStateDeleting) {
+		s.Info(fmt.Sprintf("Deleting addon %s", addonName))
+		_, err := s.ContainerEngineClient.DisableAddon(ctx, oke.DisableAddonRequest{
+			ClusterId:             okeCluster.Id,
+			AddonName:             common.String(addonName),
+			IsRemoveExistingAddOn: common.Bool(true),
+		})
+		if err != nil {
+			return err
+		}
+	} else if addonState == oke.AddonLifecycleStateDeleted {
+		s.OCIManagedControlPlane.RemoveAddonStatus(addonName)
+	}
+	return nil
+}
+
+func getAddonConfigurations(configurations []infrav2exp.AddonConfiguration) []oke.AddonConfiguration {
+	if len(configurations) == 0 {
+		return nil
+	}
+	config := make([]oke.AddonConfiguration, len(configurations))
+	for i, c := range configurations {
+		config[i] = oke.AddonConfiguration{
+			Key:   c.Key,
+			Value: c.Value,
+		}
+	}
+	return config
+}
+
+func getActualAddonConfigurations(addonConfigurations []oke.AddonConfiguration) []infrav2exp.AddonConfiguration {
+	if len(addonConfigurations) == 0 {
+		return nil
+	}
+	config := make([]infrav2exp.AddonConfiguration, len(addonConfigurations))
+	for i, c := range addonConfigurations {
+		config[i] = infrav2exp.AddonConfiguration{
+			Key:   c.Key,
+			Value: c.Value,
+		}
+	}
+	return config
+}
+
+func getAddon(addons []infrav2exp.Addon, name string) *infrav2exp.Addon {
+	for _, addon := range addons {
+		if *addon.Name == name {
+			return &addon
```

Comment:
I don't recall: do we need to do the loop variable capture stuff here? or is that only for `append` to an array?

Reply:
great point, fixed.